### ISSUE TITLE
Qz option can lead to points being lost

### ISSUE THREAD

**Issue by davidcsterratt** [qhull/qhull#25]
As part of my work on R's geometry package, I'm trying to understand how the `Qz` option works. I had been planning to make this a default option in the next release, as it is in python and octave (https://github.com/davidcsterratt/geometry/issues/4). However, it turns out that with a particular set of 2D points (attached; thanks to @Jean-Romain), this gives a triangulation with far fewer triangles than expected (https://github.com/davidcsterratt/geometry/issues/11).  

I don't think this behaviour is due to how the R geometry package has used the Qhull reentrant library, as I get similar behaviour when I give the same set of points to the qdelaunay program. Without Qz there are 156228 Delaunay regions:
```
$ bin/qdelaunay Qt < points.txt

Delaunay triangulation by the convex hull of 78150 points in 3-d:

  Number of input sites: 78142
  Number of nearly incident points: 8
  Number of Delaunay regions: 156228
  Number of triangulated facets: 7962

Statistics for:  | qdelaunay Qt

  Number of points processed: 78142
  Number of hyperplanes created: 433708
  Number of facets in hull: 156280
  Number of distance tests for qhull: 4319091
  Number of distance tests for merging: 2240387
  Number of distance tests for checking: 1180894
  Number of merged facets: 16903
  CPU seconds to compute hull (after input): 0.56
  Maximum distance of point above facet: 6.8e-09 (1.2x)
  Maximum distance of vertex below facet: -9e-09 (1.5x)
```
With Qz there are 10732 regions:
```
$ bin/qdelaunay Qt Qz < points.txt

Delaunay triangulation by the convex hull of 78151 points in 3-d:

  Number of input sites and at-infinity: 5394
  Total number of deleted points due to merging: 180
  Number of nearly incident points: 72577
  Number of Delaunay regions: 10732
  Number of triangulated facets: 3825

Statistics for:  | qdelaunay Qt Qz

  Number of points processed: 5574
  Number of hyperplanes created: 27640
  Number of facets in hull: 10784
  Number of distance tests for qhull: 10858555
  Number of distance tests for merging: 394406
  Number of distance tests for checking: 2613735
  Number of merged facets: 13954
  CPU seconds to compute hull (after input): 0.33
  Maximum distance of point above facet: 6.7e-09 (1.2x)
  Maximum distance of vertex below facet: -1.1e-08 (1.9x)
```
I'd much appreciate it if you could let me know if this behaviour is expected, and if so why.
[points.txt](https://github.com/qhull/qhull/files/984009/points.txt)

**Comment by Jean-Romain** [comment]
> However, it turns out that with a particular set of 2D points

Notice that I reproduced the bug with other datasets from laser scanning point cloud. So it is not specific to a particular set of 2D points. I can provide more examples.

**Comment by cbbarber** [comment]
This issue was transferred to
   QH6227 impossible to triangulate a dataset #12
   https://github.com/davidcsterratt/geometry/issues/12#issuecomment-400159612

It appears to be due to precision issues in the input data.   The input is integer data which loses precision due to the shared high-order digits of each coordinate.    Qhull's estimate of round-off error may be overly pessimistic in this case.    Integer data could be recognized and automatically translated to the origin.

**Comment by cbbarber** [comment]
Leaving open for further review.  The same behavior occurs with qhull 2019.1.  Notice the large number of deleted vertices and nearly incident points for 'Qz'.   These statistics indicate precision problems.    Since they only occur with 'Qz', there may be a problem with qh_partitionpoint.   I'll investigate for a later release of Qhull.

**Comment by cbbarber** [comment]
Definitely wrong.  Is this the same data as reported by Jean-Romain?  His picture clearly shows a glitch.   Looks like a bad merge or partition error.

Investigating now to fix in qhull 2019.1

**Comment by cbbarber** [comment]
On Qhull 2015.2, points.txt fails due to qh_initialhull testing for 'maybe flipped' when it should test for 'clearly flipped'.   This was previously reported, but the current code reports many coincident points.   

The new problem is that qh_scalelast ('Qbb') scaled the last coordinate to qh.MAXwide when it should have scaled to qh.MAXabs_coord.  Either choice works occur for point sets near the origin, but not for narrow point sets shifted substantially away from the origin.    This fixed the problem of many incident points.

But, without 'Qz', qhull reported cospherical data.   The problem was constructing the initial simplex from points with max or min coordinates.   If all points are tested with option 'Qs', it works OK.   Qhull 2019.1 works OK for this point set by increasing the minimum ratio for 'false_narrowhull' in qh_initialhull.  For a later release, there's a pre-existing FIX note to review the construction of the initial simplex.

Again, many thanks for reporting these problems.   I'm sorry it took so long to figure out.

**Comment by cbbarber** [comment]
Fixed in Qhull 2019.1 (not released yet).   See test runs below.

Both point sets report many Delaunay regions with co-circular point sets.    This is due to shifted data and 4-5 digits of lost precision.  If the points  for #11 are shifted by [6959, 50783], the co-circular data goes away (see below).  A future release of Qhull may automatically shift and joggle data, per vertex.

-----
qhull <../working/points-issue25.txt d Qbb

Delaunay triangulation by the convex hull of 78150 points in 3-d:

  Number of input sites: 78148
  Total number of nearly incident points: 2
  Number of Delaunay regions: 151077
  Number of non-simplicial Delaunay regions: 4860

Statistics for:  | qhull d Qbb

  Number of points processed: 78148
  Number of hyperplanes created: 437587
  Number of facets in hull: 151105
  Number of distance tests for qhull: 4281063
  Number of distance tests for merging: 2065295
  Number of distance tests for checking: 1042734
  Number of merged facets: 10255
  CPU seconds to compute hull (after input): 1.235
  Maximum distance of point above facet: 5.8e-09 (0.9x)
  Maximum distance of vertex below facet: -6.9e-09 (1.0x)

-----------------
 qhull <../working/points-issue25.txt d Qbb Qz

Delaunay triangulation by the convex hull of 78151 points in 3-d:

  Number of input sites and at-infinity: 78147
  Total number of nearly incident points: 4
  Number of Delaunay regions: 151037
  Number of non-simplicial Delaunay regions: 4891

Statistics for:  | qhull d Qbb Qz

  Number of points processed: 78147
  Number of hyperplanes created: 438334
  Number of facets in hull: 151067
  Number of distance tests for qhull: 3760884
  Number of distance tests for merging: 2065144
  Number of distance tests for checking: 1030067
  Number of merged facets: 10280
  CPU seconds to compute hull (after input): 1.195
  Maximum distance of point above facet: 6e-09 (0.9x)
  Maximum distance of vertex below facet: -9.5e-09 (1.4x)

-----
qhull <../working/dsterrat-issue11-points.txt d Qbb

Delaunay triangulation by the convex hull of 746 points in 3-d:

  Number of input sites: 723
  Total number of deleted points due to merging: 4
  Total number of nearly incident points: 19
  Number of Delaunay regions: 913
  Number of non-simplicial Delaunay regions: 384

Statistics for: davidcsterratt issue12 | qhull d Qbb

  Number of points processed: 727
  Number of hyperplanes created: 3643
  Number of facets in hull: 942
  Number of distance tests for qhull: 29202
  Number of distance tests for merging: 36478
  Number of distance tests for checking: 13726
  Number of merged facets: 1104
  CPU seconds to compute hull (after input): 0.014
  Maximum distance of point above facet: 5.8e-08 (1.0x)
  Maximum distance of vertex below facet: -9.9e-08 (1.8x)

-----
qhull <../working/dsterrat-issue11-points.txt d Qbb Qz

Delaunay triangulation by the convex hull of 747 points in 3-d:

  Number of input sites and at-infinity: 728
  Total number of deleted points due to merging: 1
  Total number of nearly incident points: 18
  Number of Delaunay regions: 923
  Number of non-simplicial Delaunay regions: 395

Statistics for: davidcsterratt issue12 | qhull d Qbb Qz

  Number of points processed: 729
  Number of hyperplanes created: 3690
  Number of facets in hull: 967
  Number of distance tests for qhull: 28136
  Number of distance tests for merging: 35430
  Number of distance tests for checking: 12858
  Number of merged facets: 1048
  CPU seconds to compute hull (after input): 0.014
  Maximum distance of point above facet: 4.3e-08 (0.8x)
  Maximum distance of vertex below facet: -1e-07 (1.8x)

-----
 qhull <../working/dsterrat-issue11-points-shifted.txt d Qbb

Delaunay triangulation by the convex hull of 746 points in 3-d:

  Number of input sites: 746
  Number of Delaunay regions: 1446

Statistics for: davidcsterratt issue12 | qhull d Qbb

  Number of points processed: 746
  Number of hyperplanes created: 4196
  Number of facets in hull: 1488
  Number of distance tests for qhull: 22218
  CPU seconds to compute hull (after input): 0.006
